File: backend/engine.py
"""
EV 충전소 수익 시뮬레이션 계산 엔진 (UI 의존성 없음)
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional


CHARGER_TYPES = {
    "완속 7kW":     {"kw": 7,   "default_rate": 320},
    "중속 30kW":    {"kw": 30,  "default_rate": 350},
    "급속 50kW":    {"kw": 50,  "default_rate": 400},
    "급속 100kW":   {"kw": 100, "default_rate": 450},
    "초급속 200kW": {"kw": 200, "default_rate": 500},
}

ELEC_PRESET = {
    "저압": {"basic": 2390,  "kwh": 128.5, "climate": 9, "fuel": -5, "fund": 2.7},
    "고압": {"basic": 2580,  "kwh": 109.5, "climate": 9, "fuel": -5, "fund": 2.7},
}


@dataclass
class ChargerConfig:
    label: str
    kw: float
    count: int
    daily_ev: float
    avg_kwh: float
    rate: float


@dataclass
class SimParams:
    charger_configs: list[ChargerConfig]
    operation_months: int = 60
    ev_growth_rate: float = 20
    rate_increase: float = 3
    elec_type: str = "저압"
    elec_basic_rate: float = 2390
    elec_kwh_rate: float = 128.5
    elec_climate_rate: float = 9
    elec_fuel_rate: float = -5
    elec_fund_pct: float = 2.7
    elec_vat_pct: float = 10.0
    cost_charger_unit: int = 1_000_000
    cost_installation: int = 2_000_000
    cost_other_init: int = 0
    monthly_ops: int = 33_000
    monthly_as: int = 0
    monthly_other: int = 0
    pg_fee_pct: float = 2.0
    revenue_share_pct: float = 90


@dataclass
class MonthRecord:
    month: int
    year_label: str
    kwh: float
    gross_revenue: float
    pg_fee: float
    net_revenue: float
    my_revenue: float
    elec_cost: float
    elec_basic: float
    elec_usage: float
    elec_fund: float
    elec_vat: float
    ops: float
    as_cost: float
    other: float
    total_cost: float
    net_profit: float
    cumulative: float


@dataclass
class SimResult:
    records: list[MonthRecord]
    total_init_cost: float
    bep_month: Optional[int]
    total_net: float
    roi: float
    avg_monthly_net: float
    kwh_total: float
    charger_summary: str


def run_simulation(p: SimParams) -> SimResult:
    total_init_cost = p.cost_charger_unit * sum(c.count for c in p.charger_configs) \
                      + p.cost_installation + p.cost_other_init

    records: list[MonthRecord] = []
    cumulative = -total_init_cost

    for m in range(1, p.operation_months + 1):
        year = (m - 1) // 12
        gf = (1 + p.ev_growth_rate / 100) ** year
        rf = (1 + p.rate_increase / 100) ** year

        gross_revenue = 0.0
        total_kwh = 0.0
        contract_kw = 0.0

        for cfg in p.charger_configs:
            t_kwh = cfg.daily_ev * cfg.count * 30 * gf * cfg.avg_kwh
            gross_revenue += t_kwh * cfg.rate * rf
            total_kwh += t_kwh
            contract_kw += cfg.kw * cfg.count

        elec_basic    = contract_kw * p.elec_basic_rate
        elec_usage    = total_kwh * (p.elec_kwh_rate + p.elec_climate_rate + p.elec_fuel_rate)
        elec_sub      = elec_basic + elec_usage
        elec_fund     = elec_sub * (p.elec_fund_pct / 100)
        elec_vat      = (elec_sub + elec_fund) * (p.elec_vat_pct / 100)
        elec_cost     = elec_sub + elec_fund + elec_vat

        pg_fee        = gross_revenue * (p.pg_fee_pct / 100)
        net_revenue   = gross_revenue - pg_fee
        my_revenue    = net_revenue * (p.revenue_share_pct / 100)
        total_cost    = elec_cost + p.monthly_ops + p.monthly_as + p.monthly_other
        net_profit    = my_revenue - total_cost
        cumulative   += net_profit

        records.append(MonthRecord(
            month=m, year_label=f"{year+1}년차",
            kwh=round(total_kwh, 1),
            gross_revenue=round(gross_revenue),
            pg_fee=round(pg_fee),
            net_revenue=round(net_revenue),
            my_revenue=round(my_revenue),
            elec_cost=round(elec_cost),
            elec_basic=round(elec_basic),
            elec_usage=round(elec_usage),
            elec_fund=round(elec_fund),
            elec_vat=round(elec_vat),
            ops=p.monthly_ops,
            as_cost=p.monthly_as,
            other=p.monthly_other,
            total_cost=round(total_cost),
            net_profit=round(net_profit),
            cumulative=round(cumulative),
        ))

    bep_month = next((r.month for r in records if r.cumulative >= 0), None)
    total_net = sum(r.net_profit for r in records) - total_init_cost
    roi = (total_net / total_init_cost * 100) if total_init_cost > 0 else 0
    avg_monthly_net = sum(r.net_profit for r in records) / len(records) if records else 0
    kwh_total = sum(r.kwh for r in records)
    charger_summary = " + ".join(f"{c.label} ×{c.count}" for c in p.charger_configs)

    return SimResult(
        records=records,
        total_init_cost=total_init_cost,
        bep_month=bep_month,
        total_net=total_net,
        roi=roi,
        avg_monthly_net=avg_monthly_net,
        kwh_total=kwh_total,
        charger_summary=charger_summary,
    )

File: frontend/src/engine.ts
import type { SimParams, SimResult, MonthRecord } from './types'

export function runSimulation(p: SimParams): SimResult {
  const totalCount = p.charger_configs.reduce((s, c) => s + c.count, 0)
  const hasPerTypeCost = p.charger_configs.some(c => c.cost_unit !== undefined)
  // 충전기 구매·설치비: 할인율 적용 대상
  const chargerCost = hasPerTypeCost
    ? p.charger_configs.reduce((s, c) => s + ((c.cost_unit ?? p.cost_charger_unit) + (c.cost_install ?? 0)) * c.count, 0)
    : (p.cost_charger_unit + p.cost_installation) * totalCount
  // 외부 고정 비용: 할인율 미적용 (한전부담금, 사용전검사비, 부속시설물)
  const fixedExtraInit = (p.cost_other_init ?? 0) + (p.cost_kepco_burden ?? 0) + (p.cost_safety_inspection ?? 0)
  const totalInitCost = chargerCost + fixedExtraInit

  const isInstallment = p.payment_type === '할부'
  const instMonths = p.operation_months
  const monthlyInstallment = isInstallment && totalInitCost > 0 && instMonths > 0
    ? totalInitCost / instMonths
    : 0

  const managerDiscount = p.manager_name ? (p.manager_discount ?? 0) : 0
  // 할인율은 충전기 비용에만 적용, 외부 고정비는 전액 반영
  const discountedInitCost = isInstallment
    ? totalInitCost
    : chargerCost * (1 - (p.discount_rate ?? 0) / 100) - managerDiscount + fixedExtraInit

  const records: MonthRecord[] = []
  let cumulative = isInstallment ? 0 : -discountedInitCost

  for (let m = 1; m <= p.operation_months; m++) {
    const year = Math.floor((m - 1) / 12)
    const gf = Math.pow(1 + p.ev_growth_rate / 100, year)
    const rf = Math.pow(1 + p.rate_increase / 100, year)

    let grossRevenue = 0
    let totalKwh = 0
    let contractKw = 0

    for (const cfg of p.charger_configs) {
      const n = Math.floor(cfg.daily_ev)
      const dailyKwh = cfg.vehicle_kwh && cfg.vehicle_kwh.length === n
        ? cfg.vehicle_kwh.reduce((s, k) => s + k, 0)
        : cfg.daily_ev * cfg.avg_kwh
      const tKwh = dailyKwh * cfg.count * 30 * gf
      grossRevenue += tKwh * cfg.rate * rf
      totalKwh += tKwh
      contractKw += cfg.kw * cfg.count
    }

    const elecBasic = contractKw * p.elec_basic_rate
    const elecUsage = totalKwh * (p.elec_kwh_rate + p.elec_climate_rate + p.elec_fuel_rate)
    const elecSub = elecBasic + elecUsage
    const elecFund = elecSub * (p.elec_fund_pct / 100)
    const elecVat = (elecSub + elecFund) * (p.elec_vat_pct / 100)
    const elecCost = elecSub + elecFund + elecVat

    const pgFee = grossRevenue * (p.pg_fee_pct / 100)
    const netRevenue = grossRevenue - pgFee
    const myRevenue = netRevenue * (p.revenue_share_pct / 100)
    const hasPerTypeOps = p.charger_configs.some(c => c.monthly_ops_unit !== undefined)
    const perTypeOps = hasPerTypeOps
      ? p.charger_configs.reduce((s, c) => {
          if (c.monthly_ops_unit === undefined) return s
          const rate = (c.count >= 2 && c.monthly_ops_bulk !== undefined) ? c.monthly_ops_bulk : c.monthly_ops_unit
          return s + rate * c.count
        }, 0)
      : p.monthly_ops
    const instThisMonth = isInstallment && m <= instMonths ? monthlyInstallment : 0
    const totalCost = elecCost + perTypeOps + p.monthly_as + (p.monthly_comm ?? 5000) + (p.monthly_elec_safety ?? 0) + (p.monthly_insurance ?? 0) + p.monthly_other + instThisMonth
    const netProfit = myRevenue - totalCost
    cumulative += netProfit

    records.push({
      month: m,
      year_label: `${year + 1}년차`,
      kwh: Math.round(totalKwh * 10) / 10,
      gross_revenue: Math.round(grossRevenue),
      pg_fee: Math.round(pgFee),
      net_revenue: Math.round(netRevenue),
      my_revenue: Math.round(myRevenue),
      elec_cost: Math.round(elecCost),
      elec_basic: Math.round(elecBasic),
      elec_usage: Math.round(elecUsage),
      elec_fund: Math.round(elecFund),
      elec_vat: Math.round(elecVat),
      ops: perTypeOps,
      as_cost: p.monthly_as,
      other: p.monthly_other,
      installment: Math.round(instThisMonth),
      total_cost: Math.round(totalCost),
      net_profit: Math.round(netProfit),
      cumulative: Math.round(cumulative),
    })
  }

  const bepMonth = records.find(r => r.cumulative >= 0)?.month ?? null
  const totalNet = records.length > 0 ? records[records.length - 1].cumulative : 0
  const roi = discountedInitCost > 0 ? (totalNet / discountedInitCost) * 100 : 0
  const avgMonthlyNet = records.length > 0
    ? records.reduce((s, r) => s + r.net_profit, 0) / records.length
    : 0
  const kwhTotal = records.reduce((s, r) => s + r.kwh, 0)
  const chargerSummary = p.charger_configs.map(c => `${c.label} ×${c.count}`).join(' + ')

  return {
    records,
    total_init_cost: discountedInitCost,
    bep_month: bepMonth,
    total_net: totalNet,
    roi,
    avg_monthly_net: avgMonthlyNet,
    kwh_total: kwhTotal,
    charger_summary: chargerSummary,
    monthly_installment: Math.round(monthlyInstallment),
  }
}
